### ISSUE TITLE
Nomos 2.6.2: output interleaved when processing directories on the commandline

### ISSUE THREAD

**Issue by armijnhemel** [fossology/fossology#396]
When I run Nomos (both nomos and nomossa) on the commandline on a directory with many files I frequently see that results of two different files are interleaved, for example:

File 1331064c29ae2a8e2d8f4fa951ff3915d7a2665736269690d384File 1ee1ab9a79f2e800a6ec644a1ddeee7ddc40ea6136dd129f1ded1000fb3f50d2 contains license(s) See-doc(OTHER)

or:

File 1a5e9eda9b465b50d4b5461a01ad899b4c5b6fc324d10a3fa9fbecca9d644e49 contaiFile 1ee1ab9a79f2e800a6ec644a1ddeee7ddc40ea6136dd
129f1ded1000fb3f50d2 contains license(s) See-doc(OTHER)

This makes it impossible for me to process (and trust!) the results from the command line when doing batch processing.

**Comment by armijnhemel** [comment]
Please note: the file names are so long and odd because I store them under the SHA256 of the content. Please do not be distracted by the file names.

I tried with both nomos and nomossa and with/without the -n option. There is no difference.
I looked at the source, but I could not quickly see how this can be fixed in a nice way.

**Comment by yaobinshi** [comment]
could anyone help to give me an example？ 
thanks.

**Comment by fogninid** [comment]
I cannot get the lines to mix up as bad as @armijnhemel,
but I get the single printfs around the `printf("File %s contains license(s) %s...` statement to interleave if I deactivate buffering:
```
$ stdbuf -o0 ./nomos -d ../agent_tests/testdata/NomosTestfiles/ -n 20
File list_sidebar.phtml contains license(s) AFL-3.0File hash_md5prime.c contains license(s) Beerware,RSA-Security

File VSL-1.0.txt contains license(s) VSL-1.0
...
```

A possible solution would be to manually buffer all the printfs and then output all at once.
Since we already have glib we can do something like:
```
gchar* toPrint;

{
 GString* str = g_string_new("");
 g_string_append_printf(str, "fmt %s", val);
 toPrint = g_string_free(str, FALSE);
}

printf("%s", toPrint);
g_free(toPrint);
```

This depends on buffer flushing setting and the printf implementation, it would be interesting to kwnow what platform @armijnhemel is using.

**Comment by bobgob** [comment]
I believe @armijnhemel is using -d so that glib buffering isn't going to work.  -d spawns a minimum of 2 nomos processes writing to stdout (more if you use -n).  The interleaving shouldn't be a problem when there is only a single process.  
A hack would be for each process to write their own output file and have the parent combine them after getting all the sigchld's.  Another solution would be to add a shared lock to gate the writes/flushes.   Each process would buffer n writes, get/wait on the lock, write/flush, release the lock and continue processing.  You wouldn't want to buffer all the writes because that could potentially eat a lot of memory.

**Comment by fogninid** [comment]
I was suggesting only buffering the output on a per scanned file basis, but I agree that this is not a complete solution and only helps to reduce the probability of getting on the bad side of the race condition.

Before implementing any complicate solution I would just use [gnu parallel](http://www.gnu.org/software/parallel/man.html#SYNOPSIS). It already implements what you are describing by buffering each process' output to a temporary file.

Small test:
```
$ time (find ../agent_tests/testdata/ -type f -print0 | parallel -0 -j8 -m 'stdbuf -o0 ./nomos'; ) > /tmp/gnuparallel

real    0m22.478s
user    1m46.094s
sys     0m4.226s
$ time (stdbuf -o0 ./nomos -d ../agent_tests/testdata/ -n 8; ) > /tmp/nomosd

real    0m20.258s
user    1m46.647s
sys     0m2.729s
$ diff <(sort /tmp/nomosd) <(sort /tmp/gnuparallel)
1d0
< 
325a325
> File gtkmovelist.c contains license(s) GPL-3.0+
495a496
> File OPL.tex contains license(s) GPL,OPL-style,Open-PL-0.4
605c606
< File StopComponentCommand.java contains license
---
> File StopComponentCommand.java contains license(s) GPL-3.0+-with-classpath-exception
```

As you see the performance hit is not so big for my machine. The extra time spent is probably going to be the same as implementing any locking solution or writing to temporary files inside nomos.

BTW, note that nomos -d lost some output there:
```
$ grep StopComponentCommand.java /tmp/nomosd -C4
File wel_image_list.h contains license(s) EFL-2.0
File serv_rssclient.c contains license(s) GPL-2.0+&GPL-3.0+
File aclocal.m4 contains license(s) FSF,GPL-2.0-with-autoconf-exception
File APSL-2.0.txt contains license(s) APSL-2.0
File StopComponentCommand.java contains license

File H2_License_1.0.rtf contains license(s) H2-1.0,Dual-license,EPL-1.0
File FSF-and-GPL.txt contains license(s) FSF,GPL-2.0+
File test2.dtd contains license(s) 3DFX
$ grep -q 'GPL-3.0+-with-classpath-exception' /tmp/nomosd || echo 'not found'
not found
$ grep -q 'gtkmovelist.c' /tmp/nomosd || echo 'not found'
not found
```

**Comment by bobgob** [comment]
Thanks @fogninid.  That's a great solution.  I wish we had thought of it before implementing -n and -d.   Given your solution and the lost output from -d, it sounds like we should just remove -n and -d. @armijnhemel what do you think?

**Comment by yaobinshi** [comment]
we need make a decision on options -d and -n.

**Comment by bobgob** [comment]
Yep.  @armijnhemel hasn't commented here so I pinged him on email.  I'm thinking that we have so many other things to do that fixing -d isn't a priority and we should just get rid of it since we have @fogninid nice shell solution.

**Comment by bobgob** [comment]
@armijnhemel pinged me offline and we both agree that working correctly is more critical than the small performance gain with -d.  So let's get rid of -d rather than invest the time into fixing it.

**Comment by mcjaeger** [comment]
so the tiket is to remove th -n and -d option?!

yaobinshi, please update assignmentin case.

**Comment by shaheemazmalmmd** [comment]
NOMOS: 4.5.0 already have fixes available for this. please reopen if some additional things needed here.
